text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+lints on dependencies are automatically capped, and the lint will not prevent
+code from compiling--and can be silenced.
+
+Expanding to include new atomics, SIMD, and other desired extensions should
+amount to a straightforward use of `cfg`.
+
+### Subsetting `std`
+
+What about subsets of `std` (or `core`)? First of all, if you apply `cfg` to
+your *crate* definition, you opt out of the default `std` portability level in
+favor of the `cfg` you write. Doing so will deny access to many APIs in `std`.
+
+Over time, APIs within `std` and `core` will be labeled with new, more narrow
+portabilities. Let's take, for example, threading--which we wish to not provide
+on platforms like Emscripten. The `std` threading APIs might be revised to use
+`cfg(threads)`, rather than `cfg(std)`; at the same time, `std` would be set up
+to imply `threads`, so that no new warnings would be generated. To check for
+compatibility with Emscripten, you can opt out of the `std` scenario, and avoid
+opting into `threads` (or use `match_cfg` if you want to do so only optionally,
+for example to use optional parallelism).
+
+Similarly, `libcore` can be annotated with new `cfg`s, like `cfg(float)` for
+floating point support.
+
+Thus, library authors shooting for maximal portability should opt out of
+`cfg(std)`, and use `cfg` as little as possible. And over time, we can allow
+increasingly fine-grained subsets of `std` by introducing new `cfg` flags.
+
+## Backwards compatibility and lint evolution
+
+The fact that the portability lint is a *lint* gives us a lot of
+flexibility. Take, for example, the assumption that `std` implies `any(windows,
+macos, linux)`. Conceivably, we may want to add more mainstream platforms in the
+future. Doing so may generate new warnings--particularly for people who had used
+`match_cfg` previously to exhaustively match against these cases. But (1) it's
+merely a new *warning*, which is fine to introduce, and can be silenced; (2)
+this is actually a highly desirable outcome if we ever did add a new mainstream
+platform, since existing code would get a heads-up that it may not longer be
+compatible with all mainstream platforms.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are several potential drawbacks to the approach of this RFC:
+
+* It adds a significant level of pedanticness about portability to Rust.
+* It does not provide airtight guarantees.
+* It may create compiler performance issues, due to the use of SAT solving.
+
+The fact that it's a lint offers some help with the first two points; the use of
+`std` as a default portability level should also help quite a bit with
+pedanticness.
+
+The worry about SAT solving is harder to mitigate; there's not much concrete
+evidence in either direction. But it is yet another place where the fact that
+it's a lint could help: we may be able to simply skip checking pathological
+cases, if they indeed arise in practice. In any case, it's hard to know how
+concerned to be until we try it.
```

Comment:
It's exponential in the number of variables, not the size of the formula (consider any formula can put into a normal form equivalent to a truth table, whose size is `2^num-vars`). Since all the variables are defined up-front, I am cautiously optimistic.